pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,282 @@
+	// among the non-exempt priority levels in proportion to their assured concurrency shares. Basically this produces
+	// the assured concurrency value (ACV) for each priority level:
+	//
+	//             ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
+	//
+	AssuredConcurrencyShares int32
+	// `queues` is a number of queues that belong to a non-exempt PriorityLevelConfiguration object. The queues exist
+	// independently at each apiserver. The value must be positive for a non-exempt priority level and setting it to 1
+	// disables shufflesharding and makes the distinguisher method irrelevant.
+	Queues int32
+	// `handSize` is a small positive number for applying shuffle sharding. When a request arrives at an apiserver the
+	// request flow identifier’s string pair is hashed and the hash value is used to shuffle the queue indices and deal
+	// a hand of the size specified here. If empty, the hand size will the be set to 1.
+	// +optional
+	HandSize *int32
+	// `queueLengthLimit` is a length limit applied to each queue belongs to the priority.  The value must be positive
+	// for a non-exempt priority level.
+	QueueLengthLimit int32
+}
+
+// PriorityLevelConfigurationConditionType is a valid value for PriorityLevelConfigurationStatusCondition.Type
+type PriorityLevelConfigurationConditionType string
+
+// PriorityLevelConfigurationStatus represents the current state of a "request-priority".
+type PriorityLevelConfigurationStatus struct {
+	// `conditions` is the current state of "request-priority".
+	Conditions []PriorityLevelConfigurationCondition
+}
+
+// PriorityLevelConfigurationCondition defines the condition of priority level.
+type PriorityLevelConfigurationCondition struct {
+	// `type` is the type of the condition.
+	Type PriorityLevelConfigurationConditionType
+	// `status` is the status of the condition.
+	// Can be True, False, Unknown.
+	Status corev1.ConditionStatus
+	// `lastTransitionTime` is the last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	// `reason` is a unique, one-word, CamelCase reason for the condition's last transition.
+	Reason string
+	// `message` is a human-readable message indicating details about last transition.
+	Message string
+}
+
+// Subject contains a reference to the object or user identities a role binding applies to.  This can either hold a direct API object reference,
+// or a value for non-objects such as user and group names.
+type Subject struct {
+	// `kind` of object being referenced. Values defined by this API group are "User", "Group", and "ServiceAccount".
+	// If the Authorizer does not recognized the kind value, the Authorizer should report an error.
+	Kind string
+	// `apiGroup` holds the API group of the referenced subject.
+	// Defaults to "" for ServiceAccount subjects.
+	// Defaults to "rbac.authorization.k8s.io" for User and Group subjects.
+	// +optional
+	APIGroup string
+	// `name` of the object being referenced.
```

Comment:
If I understand correctly, there is currently no way to write a `Subject` that matches all service accounts in a particular namespace.  Do we want to add that?

Reply:
there's a group `system:serviceaccounts` doing the same

Reply:
Oh, I see that the end of https://kubernetes.io/docs/reference/access-authn-authz/authentication/#service-account-tokens defines the group name pattern "system:serviceaccounts:(NAMESPACE)".  That accomplishes part of what am thinking about here, but not all.  Note that a service account can be matched based on its username, its group name(s), and as an object of `ServiceAccount` kind.  It is the latter case that I am thinking about here.  If we defined a value for the `Name` field here that matched all names, then this woud be another way to write a Subject that maches all service accounts in a particular namespace.  But this is not a critical need, since there is another way to do it.

For an interesting comparison, if you look at the discussion in https://groups.google.com/forum/#!topic/kubernetes-sig-auth/U_0xYKF7ldQ , you see a preference for matching service accounts as references to objects of the `ServiceAccount` kind.  The reason for that preference --- the ability to pick up the namespace from context --- does not apply here.  But it is interesting to note that there _is_ that preference for this other use of the Subject type.